documentation/Investigating-FIR-as-an-MLIR-dialect.md

```diff
@@ -0,0 +1,404 @@
+     standard tuple.
+
++    _Insert_Value_
+
+     The operation allows the insertion of a value into a composite
+     structure, such as a standard tuple.
+
++    _Field_Value_
+
+     This operation computes the offset of a component of a derived type
+     for use in ExtractValue or InsertValue operations.
+
+A number of Fortran action statements/constructs are related to
+synchronization/threading. (e.g., LOCK, UNLOCK, EVENT POST, DO
+CONCURRENT, co-arrays, etc.). The current plan is not to model these
+parallel execution semantics directly in FIR, but to lower these
+synchronization and threading statements to Fortran runtime calls. The
+implicit message passing semantics of co-arrays can be lowered to
+explicit calls as well, though the exact API is TBD.
+
+#### FIR Types
+
+Some Fortran intrinsic types are familiar and map well to MLIR standard
+types. (e.g., `INTEGER*k`, `REAL*k`, and `COMPLEX*k`.)
+However, the intrinsic
+type `CHARACTER*k(LEN=n)` has no analog. Finally, the intrinsic type
+`LOGICAL*k` and derived (user-defined) types should not be prematurely
+lowered to standard MLIR types because that may inhibit optimization and
+add complexity. (e.g., one could lower a `LOGICAL*`1 type to the standard
+`i8` type, but then it would have the same type as `INTEGER*1` and,
+depending on its usage, may have better been lowered to `i1`.)
+
+In addition to types from the surface syntax of Fortran, it is
+beneficial to introduce metatypes to FIR to capture Fortran attribute
+properties that alter the underlying object in an operational
+sense. Specifically, attributes such as `DIMENSION`, `CODIMENSION`, and
+`POINTER`, alter the size, behavior, and accepted use of a variable but
+not its type (in the Fortran sense).
+
+The additional FIR types are as follows.
+
++    __FIRCharacterType__
+
+     The Fortran intrinsic type CHARACTER with a kind value. This is
+     meant to represent the constant-sized memory reference and is
+     intentionally distinct from, for instance, an array of byte-sized
+     integers. The LEN parameter of a CHARACTER should be represented as
+     a second integer member in a FIR tuple type. Concretely, a Fortran
+     CHARACTER type is lowered into a pair:
+
+     `(FIRReferenceType<FIRCharacterType<K>>, Integer<K'>)`
+
++    __FIRLogicalType__
+
+     The Fortran intrinsic type LOGICAL with a kind value.
+
++    __FIRRealType__
+
+     The Fortran intrinsic type REAL with kind values (e.g., KIND=16)
+     that do not map to standard MLIR.
+
++    __FIRReferenceType__
+
+     The Fortran concept of "reference". Actual arguments are typically
+     passed as references to objects of some type, for example. All
+     objects that reside in memory are accessible via reference types.
+
++    __FIRSequenceType__
+
+     The Fortran concept of an object with rank > 0. Fortran does not
+     have an array type, but FIR characterizes objects with rank as
+     sequences of their base type.
```

Comment:
Does the sequence assume a row-major or column-major form for arrays?

Reply:
It's modeling Fortran explicitly, so column major.